lib/elixir/pages/Library Guidelines.md

```diff
@@ -0,0 +1,214 @@
+
+To quote [the official guide on Macros](https://elixir-lang.org/getting-started/meta/macros.html):
+
+> Even though Elixir attempts its best to provide a safe environment for macros, the major responsibility of writing clean code with macros falls on developers. Macros are harder to write than ordinary Elixir functions and it’s considered to be bad style to use them when they’re not necessary. So write macros responsibly.
+>
+> Elixir already provides mechanisms to write your everyday code in a simple and readable fashion by using its data structures and functions. Macros should only be used as a last resort. Remember that **explicit is better than implicit**. **Clear code is better than concise code**.
+
+When you absolutely have to use macro, make sure that a macro is not the only way the user can interface with your library and keep the amount of code generated by a macro to a minimum.
+
+### Avoid spawning unsupervised processes
+
+You should avoid spawning processes outside of a supervision tree, especially long-running ones. Instead, processes must be started inside supervision trees. This guarantees developers have full control over the initialization, restarts, and shutdown of the system.
+
+If your application does not have a supervision tree, one can be added by changing `def application` inside `mix.exs` to include a `:mod` key with the application callback name:
+
+```elixir
+def application do
+  [
+    extra_applications: [:logger],
+    mod: {MyApp.Application, []}
+  ]
+end
+```
+
+and then defining a `my_app/application.ex` file with the following template:
+
+```elixir
+defmodule MyApp.Application do
+    # See https://hexdocs.pm/elixir/Application.html
+    # for more information on OTP Applications
+    @moduledoc false
+
+    use Application
+
+    def start(_type, _args) do
+      # List all child processes to be supervised
+      children = [
+        # Starts a worker by calling: MyApp.Worker.start_link(arg)
+        # {MyApp.Worker, arg},
+      ]
+
+      # See https://hexdocs.pm/elixir/Supervisor.html
+      # for other strategies and supported options
+      opts = [strategy: :one_for_one, name: MyApp.Supervisor]
+      Supervisor.start_link(children, opts)
+    end
+  end
+end
+```
+
+This is the same template generated by `mix new --sup`.
+
+Each process started with the application must be listed as a child under the `Supervisor` above. We call those "static processes" because they are known upfront. For handling dynamic processes, such as the ones started during requests and other user inputs, look at the `DynamicSupervisor` module.
+
+One of the few times where it is acceptable to start a process outside of a supervision tree is with `Task.async/1` and `Task.await/2`. But this exception applies only to `Task.async/1` (and not `Task.start_link/1`) and you must always call `Task.await/2` after starting a task with `Task.async/1`.
```

Comment:
Should this suggest that, if spawning these `Task`s is a frequent thing for the app, then starting a [`Task.Supervisor`](https://hexdocs.pm/elixir/Task.Supervisor.html) upfront and using it to supervise the tasks (i.e. with something that doesn't link to the caller, like [`start_child/3`](https://hexdocs.pm/elixir/Task.Supervisor.html#start_child/3)) is another viable solution that is faster to adopt than implementing ad hoc genservers?

Reply:
> But this exception applies only to `Task.async/1` (and not `Task.start_link/1`)

It might be helpful to expand this a little. Can you make it clearer why it doesn't apply to `start_link/1`?